### ISSUE TITLE
determine when isWrapped should be reset to false

### ISSUE THREAD

**Issue by jerch** [xtermjs/xterm.js#1723]
We set `isWrapped` to `true` for a buffer line, once the text overflows right side in `wrapAround` mode in `InputHandler.print`.

We should define circumstances, when this settings has to be reset to the default value of `false`.
Currently there are a few bug reports associated with this: #1710, #1713 and a partial fix in #1720.

Proposed behavior:
- keep `isWrapped` as long as possible
- reset it to `false` once the following happens:
  - a horizontal resize event: Currently we dont support reflow on resize, therefore `isWrapped` will not reflect the correct line wraps anymore upon horizontal resize. It should simply set all `isWrapped` to `false`. Would cover #1713.
  - some cells of a line got changed by cursor repositioning: This is a tricky one since the intention of changing the old content is not clear. I think as soon as anything gets changed by repositioning the cursor the terminal is used in a more advanced manner and we should also reset `isWrapped` to `false` making them single content lines. Would cover #1710 and #1720.

Up for discussion @Tyriar, @mofux, @bgw

**Comment by jerch** [comment]
@egmontkob Thx alot, nice summary with all the bits we will need to cover this :+1:
If I get you right we would need to reset it:
- clear screen: all lines that got affected
- eraseLine
- deleteChar (somewhat surprising to me)
- any new line in `Terminal.scroll` (would also cover SU/SD)
- still unsure about resizing, I think as long as we dont support reflow resizing we might either just reset `isWrapped` and/or need some workaround in `Buffer.translateBufferLineToString` as indicated in https://github.com/xtermjs/xterm.js/issues/1713#issuecomment-425566779

About `deleteChar`: Are you doing there additional magic like rewrapping consecutive wrapped lines or is it just filling unused cells at the end of the cursor line and resetting `isWrapped`?

**Comment by egmontkob** [comment]
> If I get you right we would need to reset it:

I wouldn't use the word "need" :) I don't think there's a standard here. I'd guess it's just that xterm implemented something, then bash figured out what xterm does and hence what bash itself needs to do in order to preserve the wrapped state, and then other terminal emulators implemented something that resembles xterm enough so that maintains the wrapped state when bash edits the line (although there's no way to reasonable test all the possible ways bash can update its screen), and so on... :) My comment is not meant to describe what the "intended" behavior is, just a heads-up of what and why xterm and vte tend to do. Feel free to diverge wherever you think it makes more sense (and ideally doesn't regress bash's behavior).

> About deleteChar: Are you doing there additional magic like rewrapping

Nooooooooo!!! :) Hell I don't want to implement this :), plus I'm afraid it'd break couple of apps, especially when the number of lines decreases by one.

> just filling unused cells at the end of the cursor line and resetting isWrapped

Yup, it's that simple. This code wasn't adjusted (wasn't meant to be adjusted) for rewrap-on-resize.

**Comment by jerch** [comment]
@egmontkob Not sure if we cover the scroll region thing at that level yet, I remember having a hard to find good docs and test it reliable (ended up digging through xterm source years ago). It does something atm lol. Do you have some pointers for good doc resources beside emulator source code?

**Comment by egmontkob** [comment]
On this topic: no, not at all.

**Comment by Tyriar** [comment]
> a horizontal resize event: Currently we dont support reflow on resize, therefore isWrapped will not reflect the correct line wraps anymore upon horizontal resize. It should simply set all isWrapped to false. Would cover #1713.

I don't think we should touch isWrapped here, the fix I had in mind of #1713 was to special case it until we have reflow in place.

**Comment by jerch** [comment]
@Tyriar #1713 can be partially fixed after the pending typed array PR has landed as it also introduces a resize method at line level, which can be used to bind `line.length` to `terminal.cols` by changing `<` to `!==` here: https://github.com/xtermjs/xterm.js/blob/e7ae4be394d67ef0ca1b4f41e5251b42b9630999/src/Buffer.ts#L124 This would stop `Buffer.translateBufferLineToString` from reading beyond `terminal.cols` (which currently affects all features that rely on this method like selection and linkifier). It will not solve the `isWrapped` issue though.

**Comment by mofux** [comment]
IMO resetting `isWrapped` on horizontal resize is not correct. As far as I understand, we set `isWrapped` on lines where the text content logically belongs to the previous line (e.g. if you `cat README.md` in the xterm.js repo, any line in that file that is longer than the current `term.cols` will result in one or more additional `BufferLine`s that has `isWrapped` set to true). Since we don't automagically merge these wrapped lines with the original line as we resize the buffer horizontally, the `isWrapped` attribute is very important to keep in order to support reflow at a later stage.

**Comment by jerch** [comment]
@mofux We kinda support overlong lines only for input and for markers (includes highlighting like search/linkifier and copy stuff), but not for resize. This means if a user intends to use markers or copy&paste, he may not alter the terminal size. I think the only viable solution is to support reflow on resize, the current situation is only a poor approximation and never leads to the correct content.

To illustrate it imagine this wrapped input while cols=15:

```
buffer state:
['this is some ov']
['erlong line    ']
```
resize to cols=10:
```
buffer state:
['this is so']
['erlong lin']
```
What to do with the `isWrapped` flag? If it stays in place the buffer content would unwrap as
`'this is soerlong lin'`, thats very weird. Returning the truncated lines separately is imho closer to the real thing as it actually matches what is presented to the user.

The situation gets even more complicated since we dont really truncate the buffer lines, but "shadow" its content beyond `terminals.cols`:
```
visible:        // not visible:
['this is so']  // ['me ov']
['erlong lin']  // ['e    ']
```
Now one could argue - oh sweet the content is actually still there just not shown, lets use the whole data for the markers along with the `isWrapped` flag. Well this might work for some use cases, for others it might be seen as unexpected magical behavior. Whenever something is a matter of flavor, customizable options come in handy - but seriously, this is kinda hard to comprehend for integrators or users. Therefore I think we should not try to fix it within the current resize behavior.

TL;DR reflow resizer for the rescue, lol.

**Comment by mofux** [comment]
I see what you mean. Basically our Buffer model is in a broken state, because the buffer already makes assumptions about the terminal width. In fact, it would be much cleaner if we wouldn't restrict the BufferLine in width, but simply have one BufferLine contain all the **unwrapped** content. On top of the buffer there should then be a **view model** that transforms those raw buffer line contents into a wrapped view that gets rendered.

e.g.:

```
['this is some overlong line']
```

would end up as a single `BufferLine`, but the **view model** with 10 cols would transform it into
```
['this is so']
['me overlon']
['g line    ']
```

Is this the direction we should go for?

**Comment by jerch** [comment]
@mofux Yes this would solve those issues and even make a reflow implementation easier, and I already thought about such a radical change when I was doing the perf measuring of the typed array implementations. It also resembles parts of @Tyriar's propsal in #622, but kinda upside down.

Downside is the fact, that it will obsolete 80% of our current row/col based state handling in the core parts. Since we need those for the cursor handling we would have to calculate the correct cursor position from some kind of line stops for every action. Im sure this can be further optimized by clever stop propagation and lets say doing this kind of calculation only on the active terminal part (lines the cursor can actually reach and modify) while the scrollbuffer content is a nicely packed thing.

Imho this is definitely something we should keep in mind, not sure though if we ready for this yet.

NB: The extreme case of this would be to implement the active terminal part as a big chunk of memory with enough room for cols * rows cells (while everything else flows into the packed scrollbuffer). Not sure either if this is viable since it will introduce copy&transform actions when something flows into the scrollbuffer (seems quite expensive to me). The chunk of memory itself can be maintained as a ringbuffer with line stops. Im really curious how this would  perform, but changing 80% of the core code - well any takers?

**Comment by Tyriar** [comment]
Chatting with @jerch in Slack, I think we should just avoid fixing these issues with wrapped lines that are truncated until we do reflow, unless they're easy or someone wants to workaround them on a case-by-case basis. A proper workaround imo is not removing isWrapped but handling the invisible text as well, for example double click to select work should select inside the hidden area, clicking a link should pick up the content inside the hidden area.

But it would be a shame if these workarounds ended up taking as much time as reflow does eventually.

**Comment by jerch** [comment]
Going to close this as it is partially implemented with #1720, everything else is kinda blocked until we support the reflow resizer. Also tagging as reference since it contains valuable info from @egmontkob.
